pkg/template/params.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package template
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	pipelinev1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// AddDefaultParamsFromSpec returns the params with the addition of all
+// paramSpecs that have default values and are already in the params list
+func AddDefaultParamsFromSpecs(params []pipelinev1.Param, paramSpecs []pipelinev1.ParamSpec) []pipelinev1.Param {
+	allParamsMap := map[string]pipelinev1.ArrayOrString{}
+	for _, paramSpec := range paramSpecs {
+		if paramSpec.Default != nil {
+			allParamsMap[paramSpec.Name] = *paramSpec.Default
+		}
+	}
+	for _, param := range params {
+		allParamsMap[param.Name] = param.Value
+	}
+	allParams := []pipelinev1.Param{}
+	for name, value := range allParamsMap {
+		allParams = append(allParams, pipelinev1.Param{Name: name, Value: value})
+	}
+	return allParams
+}
+
+// ApplyParamsToResourceTemplate returns the TriggerResourceTemplate with the
+// param values substituted for all matching param variables in the template
+func ApplyParamsToResourceTemplate(params []pipelinev1.Param, rt json.RawMessage) json.RawMessage {
+	// Assume the params are valid
+	for _, param := range params {
+		rt = applyParamToResourceTemplate(param, rt)
+	}
+	return rt
+}
+
+// applyParamToResourceTemplate returns the TriggerResourceTemplate with the
+// param value substituted for all matching param variables in the template
+func applyParamToResourceTemplate(param pipelinev1.Param, rt json.RawMessage) json.RawMessage {
+	// Assume the param is valid
+	paramVariable := fmt.Sprintf("$(params.%s)", param.Name)
```

Comment:
Having the path regex equivalent of `eventPathVarRegex` would likely make this more performant/succinct.

Reply:
I don't think I understand how a regex would be better here. Are you saying it would be much faster to create a regex with the `param.Name`, then use the `regexp` package for replacing values instead of `strings.Replace` ?

Or are you saying I can organize the code more succinctly. Maybe I could turn this line into a function: `fmt.Sprintf("$(params.%s)", param.Name)` ?